Stat events

Magentah opened the pull request:
## W3C Events

Implements a `W3CEvents` Lua library that can be used to add custom events to WC3 replays.

The purpose of this is to allow map makers to add custom events that can then be used for enhanced replay analysis. This also allows W3C to add enhanced replay analysis to all melee games on the W3C ladder.

Below is an example of a `CreepKill` event that can be used to identify where the creep was killed, which unit killed it, which player killed it, and what time in game it was killed.

```json
{
  "event": "CreepKill",
  "index": 92,
  "name": "Apprentice Wizard",
  "player": 1,
  "time": 215,
  "sequence": 92,
  "killingUnit": "Blademaster",
  "x": 616.25,
  "y": 3527
}
```

## Implementation
The implementation is split across a few different files with different responsibilities

### Utilities
- `base64.lua` is a lua base64 implementation used for encoding/decoding payloads
- `debugUtils.lua` is used to print error messages and logs during games to help with debugging
- `json.lua` is used for testing and debugging by printing lua tables to json format for easier readability
- `libDeflate.lua` is a lua implementation of the deflate compression algorithm, used for compressing and decompressing payloads
- `syncedTable.lua` is a multiplayer synced table library that is not currently used.

### w3cbitbuffer.lua

This file is used to handle all reading and writing of data to payloads that will be sent using `BlzSendSyncData`. The payloads go through a series of compression steps.

- Signed integers use zigzag encoding
- Non-string values are bit-packed into a byte array
- Float values are not comressed. They're byte-aligned then written directly
- String values are byte-aligned then stored as:
    [ 2-byte big-endian length ] [ compressed bytes ]

The purpose of this is to help reduce the size of payloads to as low as possible, as the WC3 client has a hard limit of 255 bytes for `BlzSendSyncData`, and a hard limit of 4kb/s data transfer. Exceeding these values will cause high latency and potentially crash to desktop

### w3cChecksum.lua

CRC32 implementation in lua. Used to generate checksum values that are periodically sent to ensure all players in a game have the same events

### w3cdata.lua

This file is used to handle the transport layer and so implements the schema registry, compressing payloads using `w3cbitbuffer.lua`, splitting large payloads in to smaller chunked payloads, generating checksums, and decoding payloads back in to schema / values.

Schemas are used to define the fields for a singular event. Each event must correspond to a schema before it is sent.

Schemas are sent in their own payloads, allowing events themselves to only require specifying the schema id for the event. This reduces the size of each event payload. Schemas are compressed binary blobs.

COBS encoding is used to remove null bytes as `BlzSendSyncData` will truncate on null bytes

### w3cEvents.lua

The main library file that users will use. This exports a minimal set of functions required to use the library. Users will primarily use the `initialize`, `event` and `track` functions. Internally this also buffers events to reduce the number of `BlzSendSyncData` events that are sent, as well as periodically send checksum events to ensure consistent events between all players.

### w3cschema.lua

This file is used to manage the schemas that define events and their fields.

### Typescript

There is also a typescript definition file for `w3cEvents.lua`, allowing use of the library with typescript.

## Events

The MR also introduces a number of events that can be added to maps. These are all defined in the typescript files under `src/metrics`.

Magentah committed:
Initial commit for match stat events. Replays can be parsed with w3gjs but it needs to be updated to fix the 0x77 and 0x78 actions being opposite.

Parsing the replay needs to base64 decode, then zlib inflateRaw.

Magentah committed:
Updates to creating events, adding new events

marcoabreu commented on src/lua/w3cMetrics.lua:
```diff
@@ -0,0 +1,249 @@
+---@class W3Metrics
+---@field prefix string Prefix used when sending events. Used to identify syncdata packets as events. Should be 2 characters at most otherwise will cause latency issues
+---@field byteBudget integer Maximum number of bytes to send in a single SyncData packet. There's a hard limit of 255, but more than 180 will likely cause lag and desync issues. Defaults to 150.
+---@field flushInterval number How frequently to flush events in seconds. Defaults to 2.0 seconds.
```
I would recommend splitting the raw "transferring data" from the intent of the data. You're mixing communication layer (flushing, byte budgets, pooling) with application layer logic here.

If the events you're targeting here are purely for generating statistics etc, I would say that we should not have a time-based flushing interval at all. As long as the data is flushed eventually, we should not need to worry about when it's actually emitted. 

If we re-use this library for the eventual backend communication stuff, it might make sense to split the two areas.

Magentah replied:
Discussed in discord but yup, that's the plan to do before making this not a draft

Magentah replied:
Has been separated to `W3CData` that handles the data schemas, compressing, packing, chunking, etc and `W3CEvents` that handles batching, flushing, etc.

Probably some things that could be moved around but I think it's cleaner than it was before

marcoabreu commented on src/lua/w3cMetrics.lua:
```diff
@@ -0,0 +1,249 @@
+
+---comment
+---@return integer
+local function now() return math.floor(TimerGetElapsed(w3cMetrics.clock)) end
```
The game already has a clock. But it's worth noting that the game clock can be halted during lag or pauses.

Magentah replied:
From what I saw when searching, people recommended to use this approach for being able to get game time. I wasn't able to find how to get the elapsed time from the existing clock, I'm probably just blind though

marcoabreu commented on src/lua/w3cMetrics.lua:
```diff
@@ -0,0 +1,249 @@
+---@return string
+local function formatEvent(event)
+    local payload = eventPayloadString(event.value)
+    return string.format("%d|%s|%s", event.time, event.name, payload)
```
I think you need to escape the inputs if you want to use `|` as separator.

Magentah replied:
Don't think this is relevant any more

marcoabreu commented on src/lua/w3cMetrics.lua:
```diff
@@ -0,0 +1,249 @@
+    local stringEvent = formatEvent(event)
+    local compressed = LibDeflate.CompressDeflate(stringEvent);
+    local encoded = Base64.encode(compressed);
+    BlzSendSyncData("WC", encoded)
```
We should pull these constants somewhere centrally. Also, rather use a long prefix to avoid any clashes with map events.

Magentah replied:
Yup, something for me to still do.

Prefixes should limited to 2 characters otherwise there's quite significant latency impacts. no idea why but the actual values are in a forum post on hive somewhere if I remember. There's also a hard limit of 8 characters for prefixes.

marcoabreu commented on src/lua/w3cMetrics.lua:
```diff
@@ -0,0 +1,249 @@
+
+    local eventPayloads = encodePayload(w3cMetrics.events)
+    if eventPayloads then
+        for _, payload in ipairs(eventPayloads) do
```
Is it really your intent to send multiple events in one operation? I thought you were concerned about doing too much in one operation.

Magentah replied:
It's my intent to send multiple events in one send yes, as long as they fit within the size constraints of a single send. I also want to minimize the number of actual events sending as I'm somewhat concerned that one action for every event could end up causing lag/performance issues/desyncs

marcoabreu replied:
But each send call here sends an entire action, because you're actually calling SyncData. 

Didn't you want to batch multiple events into one sync data call, but spread out the sync data calls?

Calling SyncData multiple times in the same frame feels like a potential for issues.

Magentah replied:
Probably no longer relevant with the w3cdata.lua handling creating payloads to send.

marcoabreu commented on src/lua/w3cMetrics.lua:
```diff
@@ -0,0 +1,249 @@
+
+    TimerStart(timer, interval, true, function()
+        local val = nil
+        if type(getter) == "function" then val = getter() else return end
```
That's nice!

marcoabreu commented on src/lua/w3cMetrics.lua:
```diff
@@ -0,0 +1,249 @@
+---@return function stopFunction Function that can be called to stop sending the configured event.
+function w3cMetrics.track(name, getter, interval)
+    print("tracking called")
+    interval = interval or 5.0
```
I would not have a default here.

Magentah replied:
New version doesn't have any defaults for intervals

marcoabreu commented on src/metrics/heros.ts:
```diff
@@ -0,0 +1,99 @@
+        value: {},
+    }
+
+    payload.value["hero"] = GetUnitName(unit);
```
Be aware of localization and later processing - I'd recommend not working with names but IDs.

Magentah replied:
Names are mostly just for me being able to make sense of the data when testing. I also don't know if there's a way during parsing to map ids to names or if we need to extract data for that or something

marcoabreu replied:
Not sure I'm following your question. Could you rephrase?

Magentah replied:
UnitTypeIds are the id for a specific unit type, but I don't know if we have that data easily available outside of the map itself to be able to know what those ids map to when parsing the payloads later. I think that data is something we need to extract from the map files and keep as a lookup table.

marcoabreu commented on src/metrics/playerUnits.ts:
```diff
@@ -0,0 +1,54 @@
+        isHero = IsUnitType(unit, UNIT_TYPE_HERO);
+    }
+
+    const eventType =
```
I would not differentiate between heroes and units and hence emit different events but simply have the unit type be a property of the event.

Magentah replied:
Yeah, makes sense. I do that in another event I'm pretty sure. I'll need to go through and probably update the event data to be a bit nicer as they're very much just trying to see if I can get data at all and not caring about structure/schema

marcoabreu commented on src/metrics/unitDeaths.ts:
```diff
@@ -0,0 +1,104 @@
+import * as W3CMetrics from "../lua/w3cMetrics"
+import { Players } from "w3ts/globals";
+
+export function trackUnitDeaths() {
```
Be aware that custom maps do unit deaths VASTLY different. They basically never ever really kill a unit, because it leaks memory.

Not relevant for melee, but just for you to keep in mind in terms of allowing to specify custom events for "UNIT_DEATH" which are not actually EVENT_PLAYER_UNIT_DEATH. Basically to register more than 1 event that corresponds to the same thing you want to track - in melee, the size of that list would be 1 (EVENT_PLAYER_UNIT_DEATH). In customs, it might be more.

Magentah replied:
Anything in TS is only going to be used for W3C maps and not custom maps. Only the lua library would actually be used for custom maps, so I don't think it'll be a problem, but good to know!

marcoabreu replied:
Is that something we would want to change?

Of course not everything would make sense, but I think it would be good if we could have a library of defaults collectors that custom maps could use, rather than them having to define it themselves.

marcoabreu commented on src/metrics/unitDeaths.ts:
```diff
@@ -0,0 +1,104 @@
+    const killingPlayer = GetOwningPlayer(killingUnit);
+    const killingPlayerId = GetPlayerId(killingPlayer);
+
+    if (IsUnitType(killingUnit, UNIT_TYPE_HERO)) {
```
This feels .. odd. Is that the right way to do this?

Magentah replied:
No idea, I don't know anything about WC3 scripting. It works though, and at least makes sense to me that it's probably a right way to tell if a unit is a hero or not

marcoabreu replied:
Sorry, I initially had my comment at a different line. Should've been more specific. 

I mean, this whole approach of starting a timer and then reading the XP; is that the way to read the XP gained?

Magentah replied:
The timer is used here because there's a slight delay before a unit receives xp after the unit dies. If you read it immediately, the xp hasn't updated yet. I dunno if there's a better way to do this, there's no event for xp gain from what I saw when searching.

Magentah commented on src/metrics/heros.ts:
```diff
@@ -0,0 +1,99 @@
+    payload.value["skill"] = GetObjectName(GetLearnedSkill());
+    payload.value["skillLevel"] = GetLearnedSkillLevel();
+
+    W3Metrics.event("HeroSkill", payload);
```
I'm not sure about schemas tbh. I like the simplicity of basically having a simple json object with only strings/numbers. Having and enforcing schemas also makes the library itself quite a bit more complex I think, which I'd like to avoid.

As it is we can probably just let mapmakers know the schemas we expect for some things that they can use if the want, otherwise we can just parse it as keyvalues and make available for them without us providing any visualizations for them

marcoabreu replied:
Right, it's less about format and more about "these fields are expected". 

So where would the visualizations live? My understanding is that we would want to host all the visualizations, for each game mode. 

So there should be some sort of strategy as to how we deal with new or removed fields.

Magentah replied:
Probably not relevant with w3cdata.lua. I also wasn't planning on hosting any visualizations for non-melee games, I was just expecting to make the data available to map makers so they can do what they want with it.

marcoabreu commented:
We 100% need to keep batching on the communication level. We can't allow the map to just decide when it wants to emit data to our server. Remember, these are statistical events that have zero time pressure to be emitted.

Magentah committed:
Add wc3 lua definitions for lua lsp

Magentah committed:
Add new W3CData and W3CChecksum lua libraries

W3CData is used for the compression and decompression of data, as well
as the registering of schemas for event types that data is used for.

W3CChecksum is a CRC32 implementation that can be used to continuously
update a CRC value. This can then be passed to W3CData to create
checksum packets

Magentah committed:
Fix issues with packing and unpacking. Add some tests

Magentah commented:
> I am generally very impressed and satisfied with how quickly @Magentah made something quite comprehensive and mature. Obviously you're a far better and focused coder than I am, kudos to you for that.
> 
> Two things that I am concerned about here:
> 
> * encoding, as I said above, this is probably not the most effecient encoding. I would think that there's a logt of space for optimization - for example, storing time as strings is a huge waste of bandwidth, same applies to event names (as I mentioned above). I think it's important to really focus on optimizing the encoding to the maximum, because that would mean a big improvement in performance under higher load (although I have not analysed what HIGH load means in this context)

The `w3cdata.lua` should hopefully be sufficient for handling this now.

> * the batching of events - I would make it opt-in. One reason for that is, again, that you can save bandwidth, because if you send each event whenever it fires, you don't need to store the time, as you can get that from the replay then, and the time will be more reliable than a timer (although I cannot attest to how reliable or unreliable timers are).

I think this is less of a concern with w3cdata.lua. The time can be 2 bytes (for up to 65k, or 18 hours if the time is by second), or set to a specific bit length (e.g. 12 bits for up to 4k, or a bit over 1 hour). I think my approach at the moment will  likely be to buffer all events until it hits a max size (180 probably, vaguely remember seeing that number somewhere for being a soft limit) then flush, with some limit to how quickly we flush to prevent being able to send dozens of SyncData packets all at once.

Magentah committed:
Fix more issues, test overriding bit sizes

Magentah committed:
Improve input validation for integer values

Magentah committed:
Update chunking and add testing for chunking

Magentah committed:
Add W3CEvents for creating events, uses W3CData internally

Magentah committed:
Update w3cdata and w3cevents, add json.lua

W3CData:
- requires type=int to set bit sizes
- added 'number' type that allows setting minimum and maximum values
instead
- Updated checksum payloads and added test
- Added schema registry payload and added test
- Made signed default instead of unsigned
- Added function to register multiple schemas at once
- Added assertions that numbers are integers when expected

W3CEvents:
- Init now only inits once
- Added end_game that sends all remaining events and prevents any more
events. Also adds players results as the final events.
- Schemas are now required to be registered at once

json.lua added so we can easily compress and send schemas as we don't
support nested objects

Magentah committed:
Add comments to W3CData

MichaelDeitz commented:
@Magentah @marcoabreu Does this one still need to be worked on?

Magentah committed:
Updates to use w3cschema and w3cevents

Magentah committed:
Update W3CEvents defaults, add debug logs

Magentah committed:
Add order context events

Magentah committed:
Update doc comments

Magentah committed:
Fix creep kills and xp not tracking correctly

Magentah committed:
Add game end cleanup

Magentah committed:
Add additional comments

Magentah committed:
Merge Unit* and Hero* events

Magentah commented:
Think this is finally no longer a draft @marcoabreu  @diewolke9

Magentah committed:
Update flushing to be count based again

Magentah committed:
Update tracking to use a shared timer to reduce number of timers used

Magentah committed:
Add stable sorting to events using sequence numbers

Magentah committed:
Update chunks to be deterministic

Magentah committed:
Add HeroItemUse event, UPGRADE_* events, add additional data to some schemas

Magentah committed:
Remove ids causing desyncs

Magentah committed:
Revert "Remove ids causing desyncs"

This reverts commit dc18f7d67b032a812a954717c371d5759e9ce9e5.

Magentah committed:
Fix crashes caused by too many schemas registered

Magentah committed:
Add positional data to events

francislavoie commented on src/metrics/schemas.ts:
```diff
@@ -58,28 +62,48 @@ export const metricSchemas: W3CEvents.Schema[] = [
+        f("name", "string"),
+        f("killingUnit", "string"),
+        f("dyingUnitX", "float"),
+        f("dyingUnitY", "float"),
```
This should probably be `x`/`y` for consistency?

Magentah replied:
It was to make clear it's the position of the dying unit not the killing unit, but I don't mind changing it

francislavoie replied:
I think it's pretty clear given the event name. But we can just document that kind of thing I think.

Magentah committed:
Update track to use player id with event name for map

Magentah committed:
Update player sending, flushing, game ending

Player that sends events can now be configured
Player that sends events is updated when players leave
Track and Event functions no longer require playerId, expects the
playerId in event payload
Flushing is done periodically on a timer instead of based on number of
buffered events
Map game end functions now flush events first
updateMaps and build.ts now uses Wine for MPQEditor if it exists on
Linux and MacOS

Magentah committed:
Update names to use comptime generated mappings

To avoid desyncs due to different locales between players

Magentah committed:
Replace HeroDamage with CombatSummary, CombatStart and CombatEnd

Magentah committed:
Add new events, update existing events

Magentah committed:
Add worker snapshot events

Magentah committed:
Change x/y field names

Magentah committed:
Update schema registry to be compressed instead of JSON

Magentah committed:
Update CombatSummary to group for all units instead of only heroes

Magentah committed:
Fix issue with tracking events not emitting until first one in order emits

Magentah committed:
Add missing names to events with type ids

Magentah committed:
Add metrics doc for all existing metrics emitted

Magentah committed:
Fix issues with Item and Spell events

Magentah committed:
Emit checksums after flushes

Magentah committed:
Attempt to fix game end events

Magentah committed:
Update game end flushing

Magentah committed:
Cleanup W3CEvent library code
